stage0/src/initramfs.rs

```diff
@@ -16,38 +16,135 @@
 
 use crate::fw_cfg::FwCfg;
 use core::slice;
-use x86_64::VirtAddr;
+use oak_linux_boot_params::{BootE820Entry, E820EntryType};
+use x86_64::{
+    structures::paging::{PageSize, Size2MiB, Size4KiB},
+    PhysAddr, VirtAddr,
+};
 
 /// Tries to load an initial RAM disk from the QEMU FW_CFG device.
 ///
-/// If it finds a RAM disk it returns the address where it is loaded and the size. If not it
-/// returns `None`.
-pub fn try_load_initial_ram_disk(fw_cfg: &mut FwCfg) -> Option<&[u8]> {
+/// If it finds a RAM disk it returns the byte slice where it is loaded. If not it returns `None`.
+pub fn try_load_initial_ram_disk(
+    fw_cfg: &mut FwCfg,
+    e820_table: &[BootE820Entry],
+) -> Option<&'static [u8]> {
     let size = fw_cfg
         .read_initrd_size()
-        .expect("couldn't read initial RAM disk size");
+        .expect("couldn't read initial RAM disk size") as usize;
     if size == 0 {
         log::debug!("No initial RAM disk");
         return None;
     }
 
-    let address = fw_cfg
+    let mut initrd_address = fw_cfg
         .read_initrd_address()
         .expect("couldn't read initial RAM disk address");
 
-    // We use an identity mapping for memory.
-    let address = VirtAddr::new(address.as_u64());
-
     log::debug!("Initial RAM disk size {}", size);
-    log::debug!("Initial RAM disk address {:#018x}", address.as_u64());
+    log::debug!("Initial RAM disk address {:#018x}", initrd_address.as_u64());
+
+    let kernel_address = fw_cfg
```

Comment:
Two thoughts here:
a) If we want to support loading the kernel via the fw_cfg interface,  then we should mark the memory as reserved in the E820 map. (Or, as I mentioned below, check what SeaBIOS etc are doing in this case.)
b) If we load the kernel via the loader device, this address won't tell us much, as the loader has already laid out the file in memory.

tl;dr: I'm not sure how much we can trust and/or use the address and size here.

Reply:
If the kernel is an ELF file QEMU parses it and provides the already laid-out memory snapshot here, not the ELF file, and is loaded in the location specified by the ELF headers. Based on the testing I have done these values seem to match the result of loading it with a loader device.

I am currently considering also loading the kernel via the fw_cfg device, but even then I don't think we should mark it as reserved since we don't mark it as reserved when we load it via the loader device. Also, if we load a compressed image rather than an ELF file, it extracts the inner ELF and loads it elsewhere and then this memory can actually be reused.